### PR TITLE
Rough Routes and Elevations Data Retrieval

### PR DESCRIPTION
**Pull request by ufshk**
Finished:
- request of route/elevation data and response from Bing Maps API
- parsing route data  - roughly
- parsing elevation data - roughly
- main script - roughly
- writing data to csv file

Additional Notes:
- may want to add functionality of using addresses in API request, currently only coordinates work, and request functions assume user has entered coordinates
- my issues with the code from early January have been solved, they were mostly about request parameters
- Two notes on the API key
   1. Hard coding the key is probably not a good idea
   2. Right now it's the key from my account that I've connected to Bing Maps Services, it looks like they allow 50,000 transactions per day (for an education license)
API Key limitations: https://www.microsoft.com/en-us/maps/licensing/options

### CODE REVIEW

File: routemodel/maps.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+import requests
+import pandas as pd
+
+base_url = 'https://dev.virtualearth.net/REST/v1/'
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: if this is constant, should be ALL_CAPS. BASE_URL

File: routemodel/maps.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+api_key = 'An7vG9z_4w8X-6EM5WtJpP7dnIxDrZ97wLR5Ue0UAqq_aI4ZPJZPTrzNiEJj5buS'
+routes_file = 'route.csv'
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: same

File: routemodel/maps.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+elevations_csv = 'elevation.csv'
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: same

File: routemodel/maps.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+
+class MapData:
+    def __init__(self, wps: list, vwps: list, route_attrs='routePath', dist_unit='km', ht='sealevel'):
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: Docstring for the class: wps, vwps, route_attrs, dist_unit, ht

File: routemodel/maps.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        url = base_url + 'Routes?'
+        points = ''
+
+        # loop through waypoints and viawaypoints
+        # store into points string
+
+        counter = 0
+        for wp in range(len(self.waypts)):
+            for key in wp:
+                points += 'wp.' + str(counter) + '=' + str(key) + ',' \
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: base_string = 'wp.{}={},{}&.'format()

File: routemodel/maps.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                          + str(wp[key]) + '&'
+                counter += 1
+            if len(self.viawaypts[wp]) > 0:
+                for vwpkey in self.viawaypts[wp]:
+                    points += 'vwp.' + str(counter) + '=' + str(vwpkey) + ',' \
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: format

File: routemodel/maps.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        # store into points string
+
+        counter = 0
+        for wp in range(len(self.waypts)):
+            for key in wp:
+                points += 'wp.' + str(counter) + '=' + str(key) + ',' \
+                          + str(wp[key]) + '&'
+                counter += 1
+            if len(self.viawaypts[wp]) > 0:
+                for vwpkey in self.viawaypts[wp]:
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: You can use all the characters

File: routemodel/maps.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                    points += 'vwp.' + str(counter) + '=' + str(vwpkey) + ',' \
+                              + str(self.viawaypts[wp][vwpkey]) + '&'
+                    counter += 1
+
+        # add coordinates, route attribute option, distance unit, and API key
+        # to url to be requested
+        url += points + 'routeAttributes=' + self.route_attribute \
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: format

File: routemodel/maps.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        self.waypts = wps
+        self.viawaypts = vwps
+        self.route_attribute = route_attrs
+        self.distance_unit = dist_unit
+        self.heights = ht
+
+    def get_route_data(self):
+        """
+        Getting route data response from Bing Maps API.
+        @return: response from API call
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: probably want to say this is a Request.response Object

File: routemodel/maps.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        url += 'points=' + points + '&heights=' + self.heights + '&key=' + api_key
+
+        # request URL and return response
+        response = requests.get(url)
+        return response
+
+    def parse_route_data(self, response):
+        """
+        Parsing through Routes API call response.
+        @param response: API call response from get_route_data method
```
Comment (clarkevandenhoven) [inline routemodel/maps.py]: Requests object

**Comment by clarkevandenhoven** [comment]
please squash your commits

File: routemodel/maps.py (hunk 8 above)
Comment (ufshk) [reply routemodel/maps.py]: Comment resolved

File: routemodel/maps.py (hunk 7 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py (hunk 6 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py (hunk 10 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py (hunk 5 above)
Comment (ufshk) [reply routemodel/maps.py]: Kept variable as points but otherwise resolved.

File: routemodel/maps.py (hunk 3 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py (hunk 2 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py (hunk 1 above)
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+class MapData():
+    def __init__(self, wps, vwps, routeAttrs = 'routePath', distunit = 'km', ht = 'sealevel'):
+        self.waypts = wps # waypoints are what Bing Maps considers major checkpoints along a route
+        self.viawaypts = vwps # viawaypoints are what Bing Maps considers minor checkpoints along a route
+        self.route_attribute = routeAttrs # option for type of API response returned
+        self.distance_unit = distunit # unit of distance used
+        self.heights = ht # type of elevation model to be returned in API response
+
+    def GetRouteData(self):
```
[the remark this answers is not in the document]
Comment (ufshk) [reply routemodel/maps.py]: Resolved.

File: routemodel/maps.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
+import requests
+import pandas as pd
+
+base_url = 'https://dev.virtualearth.net/REST/v1/'
+api_key = 'An7vG9z_4w8X-6EM5WtJpP7dnIxDrZ97wLR5Ue0UAqq_aI4ZPJZPTrzNiEJj5buS'
```
[the remark this answers is not in the document]
Comment (ufshk) [reply routemodel/maps.py]: Resolved?

File: auxloss/auxsystem.py [hunk 13]
```diff
@@ -8,7 +8,6 @@ def __init__(self, initial_current, initial_voltage):
 
         if initial_current < 0:
             initial_current = 0
```
[the remark this answers is not in the document]
Comment (ufshk) [reply auxloss/auxsystem.py]: Resolved.

**Comment by ufshk** [comment]
ufshk does not want to merge 1 commit into uw-midsun:master from ufshk:master 👎

**Pull request closed without merging**